File: docs/gpu-market/process_gpu_data.py
#!/usr/bin/env python3
"""
显卡行情数据处理脚本
运行: python3 process_gpu_data.py [Excel文件路径]
输出:
  - data.json       : 通用JSON数据
  - src/data.js     : ES Module格式（Vite项目用）
  - excel_copy      : Excel原始文件副本（供下载）
"""

import zipfile
import xml.etree.ElementTree as ET
import json
import sys
import os
import shutil
import re

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

def parse_xlsx(filepath):
    with zipfile.ZipFile(filepath, 'r') as z:
        ss_xml = z.read('xl/sharedStrings.xml')
        ss_tree = ET.fromstring(ss_xml)
        ns = {'ns': 'http://schemas.openxmlformats.org/spreadsheetml/2006/main'}
        strings = []
        for si in ss_tree.findall('ns:si', ns):
            t = si.find('ns:t', ns)
            if t is not None:
                strings.append(t.text or '')
            else:
                # rich text with multiple runs
                runs = si.findall('ns:r', ns)
                if runs:
                    parts = [r.find('ns:t', ns) for r in runs]
                    strings.append(''.join(p.text for p in parts if p is not None))
                else:
                    # empty si
                    strings.append('')

        wb_xml = z.read('xl/workbook.xml')
        wb_tree = ET.fromstring(wb_xml)
        sheets_el = wb_tree.findall('.//{http://schemas.openxmlformats.org/spreadsheetml/2006/main}sheet')

        all_data = {}
        for i, sheet_elem in enumerate(sheets_el):
            sheet_name = sheet_elem.get('name', f'sheet{i+1}')
            sheet_xml = z.read(f'xl/worksheets/sheet{i+1}.xml')
            sheet_tree = ET.fromstring(sheet_xml)

            headers = None
            rows = []
            for row in sheet_tree.findall('.//ns:row', ns):
                row_data = []
                for cell in row.findall('ns:c', ns):
                    t = cell.get('t', '')
                    v = cell.find('ns:v', ns)
                    if v is None:
                        row_data.append('')
                    elif t == 's':
                        row_data.append(strings[int(v.text)])
                    else:
                        row_data.append(v.text)
                if headers is None:
                    headers = row_data
                else:
                    if row_data and row_data[0]:
                        rows.append(row_data)
            all_data[sheet_name] = {'headers': headers, 'rows': rows}
        return all_data


def detect_brand(name):
    n = name.lower()
    if 'nvidia' in n or n.startswith('rtx') or n.startswith('gtx'):
        return 'NVIDIA'
    elif 'amd' in n or n.startswith('rx'):
        return 'AMD'
    elif 'intel' in n or n.startswith('arc') or n.startswith('uhd'):
        return 'Intel'
    return 'Other'


def count_stars(s):
    """将 ★☆ 转换为数字 0-5"""
    if not s:
        return 0
    full = s.count('★')
    empty = s.count('☆')
    # 如果有其他字符，尝试正则
    if full == 0 and empty == 0:
        # 可能是纯数字评分
        try:
            return max(0, min(5, int(float(s))))
        except:
            return 0
    return full


def process_data(all_data):
    months = sorted(all_data.keys())
    month_labels = {
        '202512': '2025年12月', '202601': '2026年1月',
        '202602': '2026年2月', '202603': '2026年3月',
    }
    price_col_map = {
        '202512': ('11月价格', '12月价格'),
        '202601': ('12月价格', '1月价格'),
        '202602': ('1月价格', '2月价格'),
        '202603': ('2月价格', '3月价格'),
    }
    gpu_dict = {}
    base_score = 13619  # RTX 5060

    for month_key in months:
        data = all_data[month_key]
        headers = data['headers']

        def ci(name):
            try:
                return headers.index(name)
            except ValueError:
                return -1

        name_i   = ci('显卡型号')
        score_i  = ci('Time spy平均跑分')
        vram_i   = ci('显存容量')
        tdp_i    = ci('TDP功耗')
        prev_col, curr_col = price_col_map.get(month_key, (None, None))
        prev_i = ci(prev_col) if prev_col else -1
        curr_i = ci(curr_col) if curr_col else -1
        change_i = ci('涨跌幅')
        stars_i = ci('翻车概率')  # 星星列

        if name_i < 0 or score_i < 0:
            continue

        for row in data['rows']:
            if len(row) <= max(name_i, score_i, curr_i):
                continue
            name = row[name_i]
            if not name:
                continue

            score  = float(row[score_i]) if row[score_i] else 0
            vram   = row[vram_i] if vram_i < len(row) else ''
            tdp    = int(float(row[tdp_i])) if row[tdp_i] else 0
            curr   = int(float(row[curr_i])) if curr_i >= 0 and curr_i < len(row) and row[curr_i] else 0
            change = int(float(row[change_i])) if change_i >= 0 and change_i < len(row) and row[change_i] else 0
            stars_raw = row[stars_i] if stars_i >= 0 and stars_i < len(row) else ''
            stars = count_stars(stars_raw) if stars_raw else 0

            brand = detect_brand(name)
            month_label = month_labels.get(month_key, month_key)

            if name not in gpu_dict:
                gpu_dict[name] = {
                    'name': name, 'brand': brand, 'vram': vram, 'tdp': tdp,
                    'score': score, 'prices': {}, 'changes': {},
                    'performance_pct': round(score / base_score * 100, 1),
                    'stars': 0
                }
            gpu_dict[name]['prices'][month_label] = curr
            gpu_dict[name]['changes'][month_label] = change
            # 保留最高的星星数（多个月份中取最大）
            if stars > gpu_dict[name]['stars']:
                gpu_dict[name]['stars'] = stars

    for gpu in gpu_dict.values():
        latest = (gpu['prices'].get('2026年3月') or gpu['prices'].get('2026年2月') or
                  gpu['prices'].get('2026年1月') or gpu['prices'].get('2025年12月', 0))
        gpu['cost_perf'] = round(gpu['score'] / latest, 2) if latest > 0 else 0
        gpu['efficiency'] = round(gpu['score'] / gpu['tdp'], 1) if gpu['tdp'] > 0 else 0

    return list(gpu_dict.values()), months, month_labels


def generate_data_js(gpus, months, month_labels):
    """生成 ES Module 格式的 data.js"""
    src_dir = os.path.join(SCRIPT_DIR, 'src')
    os.makedirs(src_dir, exist_ok=True)

    months_out = [month_labels.get(m, m) for m in months]

    js_content = f'''// 由 process_gpu_data.py 自动生成，请勿手动修改
export const gpus = {json.dumps(gpus, ensure_ascii=False, indent=2)};
export const months = {json.dumps(months_out, ensure_ascii=False)};
'''

    path = os.path.join(src_dir, 'data.js')
    with open(path, 'w', encoding='utf-8') as f:
        f.write(js_content)
    print(f'  ✅ src/data.js 已生成')


def main():
    if len(sys.argv) >= 2:
        excel_path = sys.argv[1]
    else:
        downloads = os.path.expanduser('~/.openclaw/qqbot/downloads')
        files = [f for f in os.listdir(downloads) if f.endswith('.xlsx') and '显卡' in f]
        if not files:
            print('❌ 未找到显卡Excel文件，请指定路径: python3 process_gpu_data.py <文件路径>')
            sys.exit(1)
        files.sort(key=lambda f: os.path.getmtime(os.path.join(downloads, f)), reverse=True)
        excel_path = os.path.join(downloads, files[0])

    print(f'📄 使用文件: {excel_path}')
    all_data = parse_xlsx(excel_path)
    gpus, months, month_labels = process_data(all_data)

    # 生成 data.json（通用）
    json_path = os.path.join(SCRIPT_DIR, 'data.json')
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump({'gpus': gpus, 'months': [month_labels.get(m, m) for m in months], 'raw_months': months}, f, ensure_ascii=False, indent=2)
    print(f'  ✅ data.json 已生成 ({len(gpus)} 张显卡, {len(months)} 个月)')

    # 生成 src/data.js（Vite项目用）
    generate_data_js(gpus, months, month_labels)

    # 复制Excel
    excel_dest = os.path.join(SCRIPT_DIR, '二手显卡行情.xlsx')
    shutil.copy2(excel_path, excel_dest)
    print(f'  ✅ 原始Excel已复制到: {excel_dest}')

    print()
    print('🎉 完成！')
    print('   开发调试: npm run dev')
    print('   构建单文件: npm run build（输出 dist/index.html）')


if __name__ == '__main__':
    main()

File: projects/gpu-market/process_gpu_data.js
#!/usr/bin/env node
/**
 * 显卡行情数据处理脚本
 * 运行: node process_gpu_data.js [Excel文件路径]
 * 输出: src/data.js (ES Module)
 *
 * 特性：自动检测 Excel 中的所有月份列，无需手动配置
 * 纯 JS 实现（使用 xlsx 库）
 */

import XLSX from 'xlsx';
import { readFileSync, writeFileSync, copyFileSync, readdirSync, statSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * 从 Excel 表名解析年月，如 "202512" -> "2025年12月"
 */
function parseSheetMonth(name) {
  const m = name.trim().match(/^(\d{4})(\d{2})$/);
  if (!m) return null;
  const [, y, mo] = m;
  return `${y}年${parseInt(mo)}月`;
}

/**
 * 根据表名推导该表对应哪两个月的数据
 * 表名 202512 的数据：11月价格(上月) 和 12月价格(本月)
 * 表名 202601 的数据：12月价格(上月) 和 1月价格(本月)
 */
function deriveMonthMapping(sheetName) {
  const month = parseSheetMonth(sheetName);
  if (!month) return null;
  const m = month.match(/^(\d{4})年(\d+)月$/);
  if (!m) return null;
  const year = parseInt(m[1]);
  const monthNum = parseInt(m[2]);

  // 上个月
  let prevYear = year, prevMonth = monthNum - 1;
  if (prevMonth < 1) { prevMonth = 12; prevYear--; }
  // 本月
  let currYear = year, currMonth = monthNum;

  return {
    sheetMonth: month,
    prevMonth: `${prevMonth}月价格`,
    currMonth: `${currMonth}月价格`,
  };
}

/**
 * 检测显卡品牌
 */
function detectBrand(name) {
  const n = name.toLowerCase();
  if (n.includes('nvidia') || n.startsWith('rtx') || n.startsWith('gtx')) return 'NVIDIA';
  if (n.includes('amd') || n.startsWith('rx')) return 'AMD';
  if (n.includes('intel') || n.startsWith('arc') || n.startsWith('uhd')) return 'Intel';
  return 'Other';
}

/**
 * 计算星级（翻车概率）
 */
function countStars(s) {
  if (!s) return 0;
  const full = (s.match(/★/g) || []).length;
  if (full > 0) return Math.min(5, full);
  return 0;
}

/**
 * 解析 Excel 文件（纯 JS 实现）
 * 使用 xlsx 库读取，工作表名作为年月标识（如 "202512"）
 */
function parseXlsx(filepath) {
  const workbook = XLSX.readFile(filepath);

  // 获取所有工作表数据
  const allData = {};
  for (const sheetName of workbook.SheetNames) {
    const worksheet = workbook.Sheets[sheetName];
    // 将工作表转换为 JSON 数组（按行）
    const jsonData = XLSX.utils.sheet_to_json(worksheet, { header: 1, defval: '' });

    if (jsonData.length === 0) continue;

    // 第一行是表头
    const headers = jsonData[0].map(h => String(h).trim());
    // 后续行是数据
    const rows = jsonData.slice(1).filter(row => row.length > 0 && row[0]);

    allData[sheetName] = { headers, rows };
  }

  return allData;
}

/**
 * 主函数
 */
async function main() {
  // 获取 Excel 文件路径
  let xlsxPath = process.argv[2];
  if (!xlsxPath) {
    // 未指定参数则在项目目录下查找最新的 .xlsx 文件
    const projectDir = __dirname;
    let files;
    try {
      files = readdirSync(projectDir).filter(f => f.endsWith('.xlsx'));
    } catch (e) {
      files = [];
    }

    if (!files.length) {
      console.error('❌ 未找到 Excel 文件');
      console.error('   用法: node process_gpu_data.js <Excel文件路径>');
      console.error('   或将 Excel 文件放到项目目录下');
      process.exit(1);
    }

    // 按修改时间排序，取最新的
    files.sort((a, b) => {
      const statA = statSync(join(projectDir, a));
      const statB = statSync(join(projectDir, b));
      return statB.mtime - statA.mtime;
    });
    xlsxPath = join(projectDir, files[0]);
  }
  console.log(`📄 使用文件: ${xlsxPath}`);

  // 解析 Excel
  const allData = parseXlsx(xlsxPath);

  // 找出所有月份表（表名为纯数字年月 6 位）
  const monthSheets = Object.keys(allData).filter(k => /^\d{6}$/.test(k.trim()));
  monthSheets.sort(); // 按时间顺序排列

  if (!monthSheets.length) {
    console.error('❌ 未找到任何月份数据表（表名需为6位数字如202512）');
    process.exit(1);
  }

  console.log(`📅 检测到 ${monthSheets.length} 个月份: ${monthSheets.join(', ')}`);

  const gpuDict = {};
  const baseScore = 13619; // RTX 5060 跑分基准

  // 遍历每个月份表，提取显卡数据
  for (const mk of monthSheets) {
    const { headers, rows } = allData[mk];
    if (!headers) continue;

    // 根据表名推导月份映射
    const mapping = deriveMonthMapping(mk);
    if (!mapping) continue;

    // 查找各列索引
    const ci = n => headers.indexOf(n);
    const nameI  = ci('显卡型号');
    const scoreI = ci('Time spy平均跑分');
    const vramI  = ci('显存容量');
    const tdpI   = ci('TDP功耗');
    const prevI  = ci(mapping.prevMonth);
    const currI  = ci(mapping.currMonth);
    const changeI = ci('涨跌幅');
    const starsI  = ci('翻车概率');

    // 必需列：型号、跑分、本月价格
    if (nameI < 0 || scoreI < 0 || currI < 0) {
      console.warn(`  ⚠️ 表 ${mk} 缺少必需列，跳过`);
      continue;
    }

    for (const row of rows) {
      if (row.length <= Math.max(nameI, scoreI)) continue;
      const name = row[nameI];
      if (!name) continue;

      const score = parseFloat(row[scoreI]) || 0;
      const vram = row[vramI] || '';
      const tdp = parseInt(row[tdpI]) || 0;
      const curr = row[currI] ? parseInt(row[currI]) : 0;
      const change = changeI >= 0 && row[changeI] ? parseInt(row[changeI]) : 0;
      const stars = starsI >= 0 ? countStars(row[starsI] || '') : 0;
      const brand = detectBrand(name);

      // 首次遇到该显卡，初始化
      if (!gpuDict[name]) {
        gpuDict[name] = {
          name,
          brand,
          vram,
          tdp,
          score,
          prices: {},
          changes: {},
          performance_pct: Math.round(score / baseScore * 100 * 10) / 10,
          stars: 0,
        };
      }

      // 记录该月份的价格和涨跌幅
      gpuDict[name].prices[mapping.sheetMonth] = curr;
      gpuDict[name].changes[mapping.sheetMonth] = change;

      // 更新翻车概率（取最大值）
      if (stars > gpuDict[name].stars) {
        gpuDict[name].stars = stars;
      }
    }
  }

  const gpus = Object.values(gpuDict);

  // 从表名提取所有月份并排序
  const months = monthSheets
    .map(k => parseSheetMonth(k))
    .filter(Boolean);

  // 计算每张卡的性价比和能耗比
  for (const gpu of gpus) {
    // 取最新可用价格
    const latestPrice = [...months].reverse().find(m => gpu.prices[m]) || 0;
    gpu.cost_perf = latestPrice > 0 ? Math.round(gpu.score / latestPrice * 100) / 100 : 0;
    gpu.efficiency = gpu.tdp > 0 ? Math.round(gpu.score / gpu.tdp * 10) / 10 : 0;
  }

  // 生成 data.js 文件
  const dataJs = `// 由 process_gpu_data.js 自动生成，请勿手动修改
// 生成时间: ${new Date().toISOString()}
export const gpus = ${JSON.stringify(gpus, null, 2)};
export const months = ${JSON.stringify(months, null, 2)};
`;

  writeFileSync(join(__dirname, 'src', 'data.js'), dataJs, 'utf-8');
  console.log(`  ✅ data.js 已生成 (${gpus.length} 张显卡, ${months.length} 个月)`);

  console.log('\n🎉 完成！运行 npm run build 重新打包');
}

main().catch(e => {
  console.error(e);
  process.exit(1);
});
